numpy/core/src/umath/ufunc_object.c

```diff
@@ -2299,19 +2387,104 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
+                /* one (or more?) flexible signature names have been used. Mark them */
+                for (j = ufunc->core_offsets[i];
+                     j < ufunc->core_offsets[i] + core_num_dims[i];
+                     j++) {
+                    int dim_index = ufunc->core_dim_ixs[j];
+                    if (flexible_activated[dim_index] < 0) {
+                        flexible_activated[dim_index] = 1;
+                    }
+                    flexible_used -= flexible_activated[dim_index];
+                    if (flexible_used <= 0) {
+                        break;
+                    }
+                }
+                if (flexible_used > 0) {
+                    PyErr_Format(PyExc_ValueError,
+                         "%s: %s operand %d does not have enough "
+                         "dimensions (has %d, gufunc core with "
+                         "signature %s requires %d)",
+                         ufunc_get_name_cstr(ufunc),
+                         i < nin ? "Input" : "Output",
+                         i < nin ? i : i - nin, PyArray_NDIM(op[i]),
+                         ufunc->core_signature, n_required + flexible_used);
+                    goto fail;
+                }
+            }
+            else {
+                /* never use more than the signature requires */
+                n_dims_used[i] = core_num_dims[i];
+                /* make sure any flexible dimensions are used in the same way */
+                for (j = ufunc->core_offsets[i];
+                     j < ufunc->core_offsets[i] + core_num_dims[i];
+                     j++) {
+                    if (j <= ufunc->core_num_dim_ix) {
+                        int dim_index = ufunc->core_dim_ixs[j];
+                        if (flexible_activated[dim_index] < 0) {
+                            flexible_activated[dim_index] = 0;
+                        }
+                        else if (flexible_activated[dim_index] == 1) {
+                            PyErr_Format(PyExc_ValueError,
+                                "%s: %s operand %d with flexible dimension "
+                                "index %d signature %s previouosly marked "
+                                "flexible but now is not",
+                                 ufunc_get_name_cstr(ufunc),
+                                 i < nin ? "Input" : "Output",
+                                 i < nin ? i : i - nin, dim_index,
+                                 ufunc->core_signature);
+                            goto fail;
+                        }
+                    }
+                }
+            }
+        }
+        else {
+            /* NULL output provided. Use flexible_used to determine ndims_used */
+            n_dims_used[i] = core_num_dims[i];
+            for (j = ufunc->core_offsets[i];
+                 j < ufunc->core_offsets[i] + core_num_dims[i]; j++) {
+                if (j <= ufunc->core_num_dim_ix) {
```

Comment:
Again, I don't think this can be anything but true.

Reply:
ahh, if `keepdims` is used it can fail